### PR TITLE
feat(cancun): Establish Cancun EVM module

### LINKED ISSUE #11698: feat(cancun): Create `V050` EVM module and upgrade BESU EVM client library
Bring in BESU Cancun library (`> 24.2.1`)

- Add a `ContractsV_0_49Module` (in the mono service tree)
- Integrate it into hedera-evm service (as has been done for V038 before it)
- Test to confirm that (at least one) new Cancun opcode is accessible with that module

### LINKED ISSUE #11699: feat(cancun): HIP-865 transient storage and memory copy EVM opcode support
Support [HIP-865](https://hips.hedera.com/hip/hip-865) - "Add EVM Support for transient storage and memory copy Cancun opcodes"

* Depends on `ContractsV_0_49Module` from #11698 
* Semantics of these opcodes are _identical_ to EVM, and are fully implemented in BESU
* Provide tests that confirm these opcodes exist and run correctly in `0.49`
   * Provide tests that confirm these opcodes fail (invalid opcode) in (current) EVM module `0.38`

### LINKED ISSUE #11701: feat(cancun): HIP-867 add KZG Point Evaluation Precompiled Function support
Support [HIP-867](https://hips.hedera.com/hip/hip-867) - "Add Smart Contract Services Support for KZG Point Evaluation Precompiled Function"

* Depends on `ContractsV_0_49Module` from #11698 
* Semantics of this precompile are exactly the same as Cancun Ethereum, and is implemented by the BESU EVM client library
* Provide tests to confirm that this precompile behaves the way it should, according to the underlying EIP

### PR DESCRIPTION
**Pull request by david-bakin-sl**
**Description**:

Establish Cancun EVM module to support Ethereum Cancun EVM features as per HIPS-{865,866,867,868}.

Tests that things that are "just there" in Cancun that we get for free from BESU and use unchanged are tested.

**Related issue(s)**:

Fixes #11698
Fixes #11699 
Fixes #11701

**Notes for reviewer**:

TBD.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v049/Version049FeatureFlags.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v049/Version049FeatureFlags.java:1]: Was it necessary to create this subclass?  Could we not just use the `v046` one?

Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v049/Version049FeatureFlags.java:1]: placeholder in case we need it for something?  easier to do both these classes, side-by-side, at once?  easier for someone to read the code and see a bunch of `v049` in the source code without finding a `v046` and wondering what's up with that?

I don't feel strongly about it.  I could remove it.

File: hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+will behave consistently with the EIP-6780 `SELFDESTRUCT` rules.
+
+<!-- **TODO(Nana): Please specify which option the implementation should follow. Feel free to not that the other wasn't chosen for reason X** -->
+
+## Acceptance Tests
+
+Verify the presence of the following features in the new EVM via new smart contracts exercising their features:
+
+* `MCOPY` operation
+* `TLOAD` and `TSTORE` operations
```
Comment (lukelee-sl) [inline hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md:93]: We have custom overloaded `SLoad` and `SStore` operations.  The main reason is that we want to track the loading and storing of values into and out of storage via sidecars.  Do we not want to do this for `TLoad` and `TStore`?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v049/Version049FeatureFlags.java (hunk 1 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v049/Version049FeatureFlags.java:1]: I don't feel strongly one way or the other either, just curious.

**Review by lukelee-sl — APPROVED** [review]
LGTM.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV049.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2023-2024 Hedera Hashgraph, LLC
```
Comment (Nana-EC) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV049.java:2]: nit: 
```suggestion
 * Copyright (C) 2024 Hedera Hashgraph, LLC
```

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        assembly {
+           mcopy(0,0,32)
+        }
+    }
+
+    function try_kzg_precompile() public view {
+        bytes memory byts = "Hello, World!";
+
+        (bool success, ) = address(0x0A).staticcall(byts);
+        // If KZG precompile _succeeds_ with this bad data is must not be implemented; i.e., expect fail
```
Comment (Nana-EC) [inline hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:23]: Q: is there a simple check to confirm a success?

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        assembly {
+            tstore(10, 11111)
+            pop(tload(10))
+        }
+    }
+
+    function try_mcopy() public pure {
+        bytes memory byts = "Hello, World!";
+        assembly {
+           mcopy(0,0,32)
```
Comment (Nana-EC) [inline hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:15]: Q: is there a way to confirm the copy location?

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol [hunk 6]
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.24;
+
+contract Module049OpcodesExist {
+    function try_transient_storage() public {
+        assembly {
+            tstore(10, 11111)
+            pop(tload(10))
```
Comment (Nana-EC) [inline hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:8]: Q: is there a way to confirm the expected details popped match what was stored?

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol [hunk 7]
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.24;
+
+contract Module049OpcodesExist {
```
Comment (Nana-EC) [inline hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:4]: See https://soliditylang.org/blog/2024/01/26/transient-storage/ and https://solidity-by-example.org/transient-storage/ if it helps for coverage exploration

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV049.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV049.java:2]: Here and in a few new classes

**Review by Nana-EC — COMMENTED** [review]
Looking good, some comments and questions

File: build-logic/project-plugins/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts [hunk 8]
```diff
@@ -157,7 +157,7 @@ extraJavaModuleInfo {
         requireAllDefinedDependencies()
         requires("java.logging")
     }
-    module("com.google.guava:failureaccess", "com.google.guava.failureaccess")
+    module("com.google.guava:failureaccess", "com.google.common.util.concurrent.internal")
```
Comment (lukelee-sl) [inline build-logic/project-plugins/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts:160]: Curious why this change was necessary.

Comment (david-bakin-sl) [reply build-logic/project-plugins/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts:160]: Google changed the package and/or module that this class is in.  Got it confirmed internally (gradle slack channel) that this is necessary.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV049.java (hunk 3 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV049.java:2]: WIll fix, but spotless did that ... let me see if it does it again (if so I'll report it).

File: hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md (hunk 2 above)
Comment (david-bakin-sl) [reply hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md]: Transient storage is transient.  Disappears at the end of a tx.  Does not change persistent (blockchain) state in any way.

This is not to say it isn't interesting to a developer.  For example, a principal use case is to support reentrancy guards, as you know, something that the lack of causes contract security violations.  But (as of now) visibility into that, via the blockchain/mirror node, is not planned.

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol (hunk 4 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol]: (Discussed offline, result:) This precompile will have an additional test that checks that with a specific canned input data it actually succeeds.  Test brought over from BESU.

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol (hunk 5 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:17]: (Discussed offline, result:). These opcodes (MCOPY, TLOAD, TSTORE) are implemented by BESU itself and adopted here _without change_.  BESU has extensive tests itself (and also runs standard Ethereum EVM tests).  I consider it sufficient, here, to simply test that the opcodes exist (execute without causing an invalid operation exception) in this (Cancun) module and _don't_ exist in the previous module.

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol (hunk 6 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:10]: (see above comment at TLOAD/TSTORE.)

File: hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol (hunk 7 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/resource/contract/contracts/Module049OpcodesExist/Module049OpcodesExist.sol:6]: TY.

File: hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+3. Do not adopt but stub out placeholders. This includes the `VERSIONEDHASH` and `BLOBBASEFEE` operations.
+4. Do not adopt. This includes all Blob handling and the "Type 3" blob transaction.
+
+### Example User Story
+
+<!-- **TODO(Nana): more specific user stories highlighting the capabilities to users** -->
+
+* As a smart contract developer, I want to use current versions of solidity that may generate opcodes that are only
+  available in the Cancun fork.
+* As a Hedera developer, I want to preserve maximum future design space to adopt, adopt, or not adopt blobs.
```
Comment (IvanKavaldzhiev) [inline hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md:32]: adopt wording repetition?

Comment (david-bakin-sl) [reply hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md:32]: There's a lot of adoption going on in that sentence for sure.

Comment (david-bakin-sl) [reply hedera-node/docs/design/services/smart-contract-service/cancun-fork-support.md]: Addressed,

**Review by Nana-EC — APPROVED** [review]
LGTM

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Review by JeffreyDallas — DISMISSED** [review]

**Review by IvanKavaldzhiev — DISMISSED** [review]
LGTM

**Review by MiroslavGatsanoga — DISMISSED** [review]

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM just for the files owned by `hedera-base`. Thanks @david-bakin-sl

**Review by tinker-michaelj — APPROVED** [review]
LGTM , ty @david-bakin-sl !

**Review by jeromy-cannon — APPROVED** [review]
